daemon/container_operations.go

```diff
@@ -768,9 +768,25 @@ func (daemon *Daemon) connectToNetwork(container *container.Container, idOrName
 		container.UpdateSandboxNetworkSettings(sb)
 	}
 
-	joinOptions, err := container.BuildJoinOptions(n)
-	if err != nil {
-		return err
+	// Builds endpoint Join options from a given network.
+	var joinOptions []libnetwork.EndpointOption
+	if epConfig, ok := container.NetworkSettings.Networks[n.Name()]; ok {
+		// If Link is available, the Alias should use ShortID
+		for _, str := range epConfig.Links {
+			name, alias, err := opts.ParseLink(str)
+			if err != nil {
+				return err
+			}
+			// To preseve old behavior, the original name
+			// will be used if there is no container exists
+			if child, err := daemon.GetContainer(name); err == nil {
+				name = stringid.TruncateID(child.ID)
+			}
```

Comment:
The comment makes it look like these three lines are preserving the old behavior, but this is the implementation of the fix that turns a `name`, which coming out of `ParseLink` is really a full long ID.

I don't even remotely know how libnetwork works to suggest how the comment should be adjusted. Maybe remove the content before the comma, and leave only something like "use truncated container id if linked container was a name"